Implement PubSub api

0pg opened the pull request:
resolves #160 

Added `RedisSubscription` and `SubscriptionExecutor`

- Distinct between Publish and Subscribe commands
- Execute Subscribe commands on `SubscriptionExecutor`
- Subscribe commands with multiple parameters at once return `Stream[RedisError, (String, A)]` that tuple is (key, value) pair

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/zio/zio-redis?pullRequest=714) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/zio/zio-redis?pullRequest=714) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: 0pg<br/>:x: jahun-kim<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/zio/zio-redis?pullRequest=714) it.</sub>

0pg commented:
It might be better to get the subscription state for connection recovery 🤔

0pg commented:
@mijicd I think I've done for the first step. please give me a feedback about this PR when you are ready 😃

anatolysergeev commented on redis/src/main/scala/zio/redis/ClusterExecutor.scala:
```diff
@@ -18,18 +18,22 @@ package zio.redis
-) extends RedisExecutor {
+  scope: Scope.Closeable,
+  codec: BinaryCodec
+) extends RedisExecutor
```
Don't think that cluster executor should inherit RedisPubSub. 
I think it's more logic if it would be other way, there are would be some RedisPubSubClusterExecutor that is going to inherit ClusterExecuter, like SingleNodeRedisPubSubExecutor
And there is going to be some consistency.

mijicd commented:
@0pg please rebase and address the change requested by @anatolysergeev :pray:

0pg commented on redis/src/main/scala/zio/redis/api/PubSub.scala:
```diff
@@ -0,0 +1,89 @@
+        }
+    }
+
+  final def subscribe(channel: String, channels: String*): ZStream[Redis, RedisError, PushProtocol] =
```
> In case of multiple channels passed, we should return a List/Chunk of streams.

okay then it'd be better to get api with a single param and the callers have to handle multiple params using `ZIO.foreach`

> I think that we should hide PushProtocol from the end user.

hmm... meaning that defines separate data types containing pushed message content for only end users or registers callbacks for each pubsub command for the users who want to handle subscribe/unsubscribe messages?

0pg replied:
or using ResultBuilder that ask to give callbacks of sub/unsub when calls `returning`

anovakovic01 replied:
```scala
def subscribe(channel: String): ResultOutputStreamBuilder =
  new ResultOutputStreamBuilder(
    override def returning[R: Schema]: ZStream[Redis, RedisError, R] =
        ZStream.serviceWithStream[Redis] { redis =>
          RedisPubSubCommand
            .run(RedisPubSubCommand.Subscribe(channel))
            .collect { case t: PushProtocol.Message => t.message }
            .mapZIO { resp =>
              ZIO
                .attempt(ArbitraryOutput[R]().unsafeDecode(resp)(redis.codec))
                .refineToOrDie[RedisError]
            }
        }
  )

def subscribe(channels: Chunk[String]): ResultOutputMultiStreamBuilder =
  new ResultOutputMultiStreamBuilder(
    override def returning[R: Schema]: UIO[Chunk[ZStream[Redis, RedisError, R]]] =
        ZIO.serviceWithZIO[Redis] { redis =>
          RedisPubSubCommand
            .run(RedisPubSubCommand.Subscribe(channels.head, channels.tail))
            .collect { case t: PushProtocol.Message => t.message }
            .broadcast(channels.length, maximumLag)
            .map(streams =>
              streams.zipWith(channels) { (stream, channel) =>
                stream
                  .filter(_.key.value == channel)
                  .mapZIO { resp =>
                    ZIO
                      .attempt(ArbitraryOutput[R]().unsafeDecode(resp)(redis.codec))
                      .refineToOrDie[RedisError]
                  }
              }
            )
        }
```
This is just a sketch. I hope that I managed to explain what's the base idea through this code. Of course, something like `ResultOutputMultiStreamBuilder` should be added for multi-stream scenarios.

jahun-kim replied:
yeah that's almost same what I though and mentioned above 👍

0pg commented:
@anovakovic01 Hi there are some changes please check 🙏 
- added `ResultStreamBuilder[+F[_]`
- added subscription callback

Cluster PubSubExecutor will be worked by another PR

anovakovic01 commented on redis/src/main/scala/zio/redis/api/PubSub.scala:
```diff
@@ -0,0 +1,158 @@
+
+  final def subscribeWithCallback(
+    channel: String,
+    channels: List[String]
```
You can use `String*` here.

anovakovic01 commented on redis/src/main/scala/zio/redis/api/PubSub.scala:
```diff
@@ -0,0 +1,158 @@
+  final def subscribe(channel: String): ResultStreamBuilder[Id] =
+    subscribeWithCallback(channel)(emptyCallback)
+
+  final def subscribe(channel: String, channels: List[String]): ResultStreamBuilder[List] =
```
You can use `String*` here.

anovakovic01 commented on redis/src/main/scala/zio/redis/api/PubSub.scala:
```diff
@@ -0,0 +1,158 @@
+    RedisPubSubCommand(PubSubCommand.Subscribe(channel, channels, onSubscribe), codec, pubSub).run
+      .map(_.map(_.mapZIO {
+        case PushProtocol.Message(_, msg) =>
+          ZIO
+            .attempt(ArbitraryOutput[R]().unsafeDecode(msg)(codec))
+            .refineToOrDie[RedisError]
+            .asSome
```
I think that this can be part of the `run` implementation in the `RedisPubSubCommand`.

0pg replied:
I made `RedisPubSubCommand` takes responsibilities of transforming message types and invoking callback

anovakovic01 commented on redis/src/main/scala/zio/redis/PubSubCommand.scala:
```diff
@@ -0,0 +1,29 @@
+  }
+}
+
+sealed trait PubSubCommand
+
+object PubSubCommand {
+  case class Subscribe(
+    channel: String,
+    channels: List[String],
+    onSubscribe: PubSubCallback
+  ) extends PubSubCommand
+  case class PSubscribe(
+    pattern: String,
+    patterns: List[String],
+    onSubscribe: PubSubCallback
+  ) extends PubSubCommand
+  case class Unsubscribe(channels: List[String])  extends PubSubCommand
+  case class PUnsubscribe(patterns: List[String]) extends PubSubCommand
+}
```
We should hide these from the end user. These should be in a dedicated `PubSubCommand` file.

0pg replied:
I moved it to its own file but do we need to add package private access modifier?

anovakovic01 replied:
I think that we do if it's not part of the `api` interface.

0pg replied:
I added package private accessor

0pg commented on redis/src/main/scala/zio/redis/options/PubSub.scala:
```diff
@@ -0,0 +1,51 @@
+    case class Pattern(value: String) extends SubscriptionKey
+  }
+
+  case class NumSubResponse(channel: String, subscriberCount: Long)
```
I renamed and moved it inside to companion object of `options/PubSub`

0pg commented on redis/src/main/scala/zio/redis/options/PubSub.scala:
```diff
@@ -0,0 +1,51 @@
+
+  object PushProtocol {
+    case class Subscribe(channel: String, numOfSubscription: Long) extends PushProtocol {
+      def key: SubscriptionKey = SubscriptionKey.Channel(channel)
```
you mean that gets rid of `SubscriptionKey`? or changes return type of `def key` to `String`?
I changed return type to `String`

anovakovic01 replied:
Sorry for not being clear. I meant to remove `SubscriptionKey` entirely and replace it with a plain `String`.

0pg replied:
ah I moved `SubscriptionKey` into `SingleNodeRedisPubSub` as a private case class

0pg commented:
@anovakovic01 please review when you are free 🙏 
- Remove `PushProtocol` from `RedisPubSubCommand`'s interface
- Some changes in `SingleNodeRedisPubSub` to guarantee delivering subscribe/unsubscribe response
- Using `Take[RedisError, PushProtocol]` to release stream

anovakovic01 commented on redis/src/main/scala/zio/redis/RedisPubSubCommand.scala:
```diff
@@ -0,0 +1,25 @@
+  def run[R: Schema](): IO[RedisError, Chunk[Stream[RedisError, R]]] = {
+    val codecLayer = ZLayer.succeed(codec)
+    executor
+      .execute(command)
```
Can you do the encoding here and avoid passing the codec as a layer?

0pg replied:
it can be removed from `RedisPubSub` interface but `SignleNodeRedisPubSub` needs codec to managing sub/unsub state and re-subscribe for failover so I moved codec into `SignleNodeRedisPubSub`'s field

anatolysergeev commented on redis/src/main/scala/zio/redis/ClusterExecutor.scala:
```diff
@@ -150,9 +150,11 @@ object ClusterExecutor {
+    val codecLayer       = ZLayer.succeed[BinaryCodec](StringUtf8Codec)
+    val executorLayer    = redisConfigLayer >>> RedisExecutor.layer
+    val pubSubLayer      = redisConfigLayer ++ codecLayer >>> RedisPubSub.layer
+    val redisLayer       = executorLayer ++ pubSubLayer ++ codecLayer >>> Redis.layer
```
🤔 
Strange that to execute redis simple command we need pub sub layer.
Should it be part of Redis.Live ?

0pg replied:
@anatolysergeev I separated subscription part of PubSub into `RedisSubscription`
it includes only commands (`subscribe`, `psubscribe`, `unsubscribe`, `punsubscribbe`) that changes protocol resp to pushed type.

anovakovic01 commented on redis/src/main/scala/zio/redis/SingleNodeRedisPubSub.scala:
```diff
@@ -0,0 +1,250 @@
+  )(implicit codec: BinaryCodec) =
+    for {
+      promises <- Promise.make[RedisError, PushProtocol].replicateZIO(keys.size + 1).map(Chunk.fromIterable(_))
+      chunk     = StringInput.encode(command) ++ NonEmptyList(StringInput).encode((key.value, keys.map(_.value)))
```
Can this be done in the `RedisPubSubCommand`?

anovakovic01 reviewed (commented):
Please remove an empty `RedisExecutor` file. @mijicd Can you also take a look at this PR? I'm not sure what to do about an executor.

anovakovic01 commented on modules/redis/src/main/scala/zio/redis/Output.scala:
```diff
@@ -639,6 +640,50 @@ object Output {
+        case RespValue.Array(values) =>
+          Chunk.fromIterator(values.grouped(2).map { chunk =>
+            val channel           = MultiStringOutput.unsafeDecode(chunk(0))
+            val numOfSubscription = LongOutput.unsafeDecode(chunk(1))
```
You can rename this to `numOfSubs`

anovakovic01 commented on modules/redis/src/main/scala/zio/redis/options/PubSub.scala:
```diff
@@ -0,0 +1,20 @@
+
+import zio.UIO
+import zio.redis.internal.RespValue
+object PubSub {
```
Add an empty line between the import block and the object.

0pg replied:
I added https://github.com/zio/zio-redis/pull/714/commits/1a0862f7b30f2dc8a611217c98e8f85814276f2b

0pg commented:
@anovakovic01 I fixed some cross compile issues 🙏 
https://github.com/zio/zio-redis/pull/714/commits/92457c4f59bfd12b1f51d31fab3bad38aab34fba

mijicd commented:
@anovakovic01 this PR needs to address a few things before merging (compilation, linting, sync with master)

jahun-kim commented:
Is there anything to do about test failures? it seems like non deterministic 🤔

0pg commented:
Is there anything to do about test failures? it seems like non deterministic 🤔

mijicd commented:
It looks like these aren't directly related to your change, so I'd say let's not touch any of them in this PR.

mijicd commented on modules/redis/src/main/scala/zio/redis/RedisSubscription.scala:
```diff
@@ -0,0 +1,34 @@
+
+import zio._
+
+trait RedisSubscription extends api.Subscription
```
Just to clarify, are you using the `RedisSubscription` to ensure clients cannot call other methods while in the subscription regime?

0pg replied:
That was my intended design. However, the current implementation does not block other commands, but instead requires access through a subscription API to interact with the implementation.

mijicd commented on modules/redis/src/main/scala/zio/redis/SingleNodeSubscriptionExecutor.scala:
```diff
@@ -0,0 +1,219 @@
+ * limitations under the License.
+ */
+
+package zio.redis
```
It belongs to `zio.redis.internal`.

mijicd commented on modules/redis/src/main/scala/zio/redis/Output.scala:
```diff
@@ -639,6 +640,50 @@ object Output {
+              val num = LongOutput.unsafeDecode(values(2))
+              PushProtocol.PUnsubscribe(key, num)
+            case "message" =>
+              PushProtocol.Message(key, values(2))
```
I would assign `values(2)` to a val.

mijicd commented on modules/redis/src/main/scala/zio/redis/SingleNodeSubscriptionExecutor.scala:
```diff
@@ -0,0 +1,219 @@
+        .tapError(e => ZIO.logError(s"Executor exiting: $e"))
+}
+
+private[redis] object SingleNodeSubscriptionExecutor {
```
This whole module if full of copy-paste. Let's try to reduce that.

0pg replied:
I made `SingleNodeRunner` trait that has two abstract functions `send`, `receive` to extract `run` function

mijicd commented on modules/redis/src/main/scala/zio/redis/SingleNodeSubscriptionExecutor.scala:
```diff
@@ -0,0 +1,219 @@
+import zio.{Chunk, ChunkBuilder, IO, Promise, Queue, Ref, Schedule, Scope, URIO, ZIO}
+
+private[redis] final class SingleNodeSubscriptionExecutor private (
+  channelSubsRef: Ref[Map[String, Chunk[Queue[Take[RedisError, PushProtocol]]]]],
+  patternSubsRef: Ref[Map[String, Chunk[Queue[Take[RedisError, PushProtocol]]]]],
```
Can't we use hubs?

0pg replied:
This is to ensure the callback is invoked. response messages can be delivered from Redis before target streams are consumed. if we use hub then response messages can be loss that have to invoke callbacks

0pg replied:
but maybe It seems like we could use a promise queue for handling response messages (like `SingleNodeExecutor`) to enable using the hub 🤔

0pg replied:
I added promise queue for waiting subscribe responses and let them use `Hub` intead of chunks of `Queue`

mijicd commented on modules/redis/src/main/scala/zio/redis/SubscriptionExecutor.scala:
```diff
@@ -0,0 +1,42 @@
+import zio.stream._
+import zio.{Layer, ZIO, ZLayer}
+
+trait SubscriptionExecutor {
```
It doesn't belong here. Similarly to all other executors, it should go to `zio.redis.internal`.

mijicd commented on modules/redis/src/main/scala/zio/redis/SubscriptionExecutor.scala:
```diff
@@ -0,0 +1,42 @@
+
+object SubscriptionExecutor {
+  lazy val layer: ZLayer[RedisConfig, RedisError.IOError, SubscriptionExecutor] =
+    RedisConnection.layer.fresh >>> pubSublayer
```
Why do you need `fresh`?

0pg replied:
As far as I know, when a client subscribes to channels, it is restricted from sending most commands except for a few that are allowed in a subscribing connection

mijicd commented on modules/redis/src/main/scala/zio/redis/SubscriptionExecutor.scala:
```diff
@@ -0,0 +1,42 @@
+    RedisConnection.layer.fresh >>> pubSublayer
+
+  lazy val local: Layer[RedisError.IOError, SubscriptionExecutor] =
+    RedisConnection.local.fresh >>> pubSublayer
```
Same here.

mijicd commented on modules/redis/src/main/scala/zio/redis/api/Subscription.scala:
```diff
@@ -0,0 +1,124 @@
+    new ResultStreamBuilder1[({ type lambda[x] = (String, x) })#lambda] {
+      def returning[R: Schema]: Stream[RedisError, (String, R)] =
+        RedisSubscriptionCommand(executor).subscribe(
+          Chunk.single(channel) ++ Chunk.fromIterable(channels),
```
No need for chunk concatenation.

mijicd commented on modules/redis/src/main/scala/zio/redis/api/Subscription.scala:
```diff
@@ -0,0 +1,124 @@
+    }
+
+  final def pSubscribe(
+    pattern: String,
+    patterns: String*
```
Type-safety could be improved.

mijicd commented on modules/redis/src/main/scala/zio/redis/api/Subscription.scala:
```diff
@@ -0,0 +1,124 @@
+}
+
+object Subscription {
+  private lazy val emptyCallback = (_: String, _: Long) => ZIO.unit
```
Let's follow the constants definition guideline. Also no need to make it lazy.

mijicd commented on modules/redis/src/main/scala/zio/redis/internal/SubscribeEnvironment.scala:
```diff
@@ -0,0 +1,27 @@
+import zio.schema.Schema
+import zio.schema.codec.BinaryCodec
+
+private[redis] trait SubscribeEnvironment {
```
We might be able to generalize the environments, though I'm fine with exploring that possibility in a follow-up.

mijicd commented on modules/redis/src/test/scala/zio/redis/OutputSpec.scala:
```diff
@@ -1,9 +1,10 @@
 
 import zio._
-import zio.redis.Output._
+import zio.redis.Output.{PushProtocolOutput, _}
```
This is an unnecessary change.

mijicd reviewed (changes requested):
Thank you for taking care of this. It's a massive chunk of work, and it's going in the right direction. 

During the first review pass, I focused on the package layout, top-level abstractions, and protocol correctness. Once all of them are in place, I'll do a more detailed review of its internals, specifically the streaming part.

One general remark is that we can improve the test suite to use property tests and cover more than just a "happy path". Note that this remark applies to most tests, but we must start improving their state somewhere.

0pg commented on modules/redis/src/main/scala/zio/redis/options/PubSub.scala:
```diff
@@ -0,0 +1,37 @@
+    case class Unsubscribe(channel: String, numOfSubs: Long)                  extends PushProtocol
+    case class PUnsubscribe(pattern: String, numOfSubs: Long)                 extends PushProtocol
+    case class Message(channel: String, message: RespValue)                   extends PushProtocol
+    case class PMessage(pattern: String, channel: String, message: RespValue) extends PushProtocol
```
it also changes in `SingleNodeSubscriptionExecutor`'s member `Ref` that manages subscription state

0pg commented on modules/redis/src/main/scala/zio/redis/Output.scala:
```diff
@@ -639,6 +640,50 @@ object Output {
       }
   }
 
+  case object PushProtocolOutput extends Output[PushProtocol] {
```
`PushMessageOutput` is only used in `internal` package so I added `private[redis]` to this object
https://github.com/zio/zio-redis/pull/714/commits/a1546838ca3ed7a8b88be844069c489ec5943a19

0pg commented:
@mijicd Thank you for your review. I worked some parts based on comments. please to check changes and reply comments when you're fine. 😃

mijicd commented on modules/redis/src/main/scala/zio/redis/internal/SingleNodeSubscriptionExecutor.scala:
```diff
@@ -0,0 +1,218 @@
+    def getHub(key: SubscriptionKey): IO[RedisError, Hub[Take[RedisError, PushMessage]]] =
+      subsRef.get
+        .map(_.get(key))
+        .flatMap(ZIO.fromOption(_))
```
I think you can go without `(_)`. Do that on all applicable places.

mijicd reviewed (commented):
@anovakovic01 @anatolysergeev Please take a look!

0pg commented on modules/redis/src/main/scala/zio/redis/internal/PubSub.scala:
```diff
@@ -0,0 +1,38 @@
+
+package zio.redis.internal
+
+object PubSub {
```
I marked it to them

0pg commented on modules/redis/src/main/scala/zio/redis/internal/RequestQueue.scala:
```diff
@@ -0,0 +1,24 @@
+
+private[redis] object RequestQueue {
+  private final val RequestQueueSize = 16
+  def create[A]: UIO[Queue[A]]       = Queue.bounded[A](RequestQueueSize)
```
I moved constant to package object and removed this object

0pg commented on modules/redis/src/main/scala/zio/redis/internal/SingleNodeRunner.scala:
```diff
@@ -0,0 +1,44 @@
+      (send.repeat(Schedule.forever) race receive)
+        .tapError(e => ZIO.logWarning(s"Reconnecting due to error: $e") *> onError(e))
+        .retryWhile(True)
+        .tapError(e => ZIO.logError(s"Executor exiting: $e"))
```
I marked it as private final

anovakovic01 commented on modules/redis/src/main/scala/zio/redis/internal/SingleNodeSubscriptionExecutor.scala:
```diff
@@ -0,0 +1,221 @@
+    }.map(SubscriptionKey.Pattern.apply)
+
+  def send: IO[RedisError.IOError, Unit] =
+    requests.takeAll.flatMap { reqs =>
```
Use `takeBetween(1, RequestQueueSize)` just like in the existing `send` [implementation](https://github.com/zio/zio-redis/blob/master/modules/redis/src/main/scala/zio/redis/internal/SingleNodeExecutor.scala#L49).

anovakovic01 commented on modules/redis/src/main/scala/zio/redis/api/Subscription.scala:
```diff
@@ -0,0 +1,123 @@
+
+trait Subscription extends SubscribeEnvironment {
+
+  final def subscribe(channel: String): ResultStreamBuilder1[Id] =
```
Add doc comments for the public API.

0pg replied:
I added doc `Subscription` and `Publishing` apis

anovakovic01 commented:
@0pg Great work! You're almost there, these are some minor changes that I've commented on, but nothing major.

0pg commented on modules/redis/src/main/scala/zio/redis/api/Subscription.scala:
```diff
@@ -0,0 +1,123 @@
+
+  final def subscribeWithCallback(channel: String)(
+    onSubscribe: Option[PubSubCallback],
+    onUnsubscribe: Option[PubSubCallback]
```
yeah it seems to make more ergonomic api. I changed to use `NoopCallback`. 
I changed function name `subscribe` to `subscribeSingle` for only single channel subscription call because it's more clear to distinct multiple channels subscription call (also they have different return type signature)
[`5ca98e7` (#714)](https://github.com/zio/zio-redis/pull/714/commits/5ca98e7d763e1b85998623aeabf4935f8fe69405)

0pg commented on modules/redis/src/main/scala/zio/redis/internal/SingleNodeSubscriptionExecutor.scala:
```diff
@@ -0,0 +1,221 @@
+                         .fromOption(command.args.collectFirst { case RespCommandArgument.CommandName(name) =>
+                           name
+                         })
+                         .orElseFail(RedisError.CommandNameNotFound(command))
```
Thrown `ProtocolError` more makes sense 🤔 I'll try to this using `NonEmptyChunk` approach in another PR

0pg replied:
I made PR that gives name separated field
https://github.com/zio/zio-redis/pull/852
